packages/dd-trace/src/appsec/iast/vulnerability-reporter.js

```diff
@@ -15,39 +16,59 @@ const RESET_VULNERABILITY_CACHE_INTERVAL = 60 * 60 * 1000 // 1 hour
 let tracer
 let resetVulnerabilityCacheTimer
 let deduplicationEnabled = true
+let stackTraceEnabled = true
+let stackTraceMaxDepth
+let maxStackTraces
 
-function addVulnerability (iastContext, vulnerability) {
-  if (vulnerability?.evidence && vulnerability?.type && vulnerability?.location) {
-    if (deduplicationEnabled && isDuplicatedVulnerability(vulnerability)) return
+function canAddVulnerability (vulnerability) {
+  const hasRequiredFields = vulnerability?.evidence && vulnerability?.type && vulnerability?.location
+  if (!hasRequiredFields) return false
 
-    VULNERABILITY_HASHES.set(`${vulnerability.type}${vulnerability.hash}`, true)
+  const isDuplicated = deduplicationEnabled && isDuplicatedVulnerability(vulnerability)
 
-    let span = iastContext?.rootSpan
+  return !isDuplicated
+}
 
-    if (!span && tracer) {
-      span = tracer.startSpan('vulnerability', {
-        type: 'vulnerability'
-      })
+function addVulnerability (iastContext, vulnerability, callSiteFrames, stackId) {
+  if (!canAddVulnerability(vulnerability)) return
 
-      vulnerability.location.spanId = span.context().toSpanId()
+  VULNERABILITY_HASHES.set(`${vulnerability.type}${vulnerability.hash}`, true)
 
-      span.addTags({
-        [IAST_ENABLED_TAG_KEY]: 1
-      })
-    }
+  let span = iastContext?.rootSpan
 
-    if (!span) return
+  if (!span && tracer) {
+    span = tracer.startSpan('vulnerability', {
+      type: 'vulnerability'
+    })
 
-    keepTrace(span, SAMPLING_MECHANISM_APPSEC)
-    standalone.sample(span)
+    vulnerability.location.spanId = span.context().toSpanId()
 
-    if (iastContext?.rootSpan) {
-      iastContext[VULNERABILITIES_KEY] = iastContext[VULNERABILITIES_KEY] || []
-      iastContext[VULNERABILITIES_KEY].push(vulnerability)
-    } else {
-      sendVulnerabilities([vulnerability], span)
-      span.finish()
-    }
+    span.addTags({
+      [IAST_ENABLED_TAG_KEY]: 1
+    })
+  }
+
+  if (!span) return
+
+  keepTrace(span, SAMPLING_MECHANISM_APPSEC)
+  standalone.sample(span)
+
+  if (stackTraceEnabled) {
+    reportStackTrace(
+      iastContext?.rootSpan,
```

Comment:
if there is no `rootSpan` a new one is created to report the vulnerability. Are we not interested in the stackTrace in that case?

```suggestion
      span,
```

Reply:
Yes we should use span if there is no rootSpan, will fix it